Epic Labs verification classifier integration.

j0sh opened the pull request:
**What does this pull request do? Explain your changes. (required)**

Extremely early integration of the Epic Labs verification classifier API. This should be enough to start running some experiments locally.

Refer to https://github.com/livepeer/go-livepeer/issues/1093 for some details on the following items:

 - Due to https://github.com/livepeer/verification-classifier/issues/64 this works when the verifier is running on localhost only. There is a lot hard-coded in here.
- Results (and errors) from the verification function are ignored completely.
- Adds a new CLI flag, `-verifierUrl` (but this may change in later iterations of the integration)
- Only performs verification if the `-verifierUrl` address is set.
- Currently works in off-chain mode to facilitate quick testing.
- Verifies all segments and renditions; no sampling is done.
- Structure is subject to change as more of the policy is designed and implemented.
- Leaves surrounding code intact, eg pixel counting and sig checking.

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->
- 
- 
- 

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->


**Does this pull request close any open issues?**
<!-- Fixes # -->


**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

j0sh committed:
epic: Simple integration with the Epic Labs verification classifier.

j0sh committed:
cmd: Use CLI flag to indicate verifier address.

* Make the use of the verifier conditional on this address.

Also define an early interface for the verifier.

j0sh committed:
server, verification: Add tests.

j0sh committed:
epic: Process API results.

j0sh committed:
verification: Policy implementation.

j0sh committed:
verification: Use pixel checking results from verifier.

If the verifier itself is available.

yondonfu commented on server/broadcast.go:
```diff
@@ -434,6 +444,13 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string)
 		}
+
+		if verifier != nil {
+			if err := verify(verifier, cxn, sess, seg, res); err != nil {
```
If we end up initializing a [verifier-less policy](https://github.com/livepeer/go-livepeer/pull/1096/files#diff-205b550ef0ac99b368da16cfb883960bR664) responsible for sig & pixel count checking when running in on-chain mode, then pixel count checking would become synchronous instead of asynchronous. That might be an ok tradeoff for cleaner separation of verification related concerns - just noting this for now.

yondonfu commented on server/broadcast.go:
```diff
@@ -462,6 +479,35 @@ func shouldStopSession(err error) bool {
+	// Verify. The params we receive, if any, are the *accepted* params
+	// The accepted params are not necessarily the same params we just have!
+	p, err := verifier.Verify(params)
+	if _, retry := err.(verification.Retryable); retry {
```
Is the plan to just check for `verification.Retryable` errors for now and consider distinguishing between fatal and verifier related errors later (as described in #1100)?

j0sh replied:
Good point, I missed that the spec distinguishes between verifier errors and other types of fatal verification-related errors. I had been lumping them all under "retryable" errors (called such because a little weird to be retrying on supposed "fatal" errors).

The spec does imply that verifier errors shouldn't immediately remove the orchestrator from the working set, while fatal errors should - which is probably sensible and the reason for making a distinction between the two. Will adjust.

yondonfu commented on verification/verify.go:
```diff
@@ -0,0 +1,133 @@
+	Results *net.TranscodeData
+}
+
+type VerificationResult interface {
```
Why use an interface here? Do you imagine implementations that do more than just return internal score and pixels fields (as is done by the `verificationResult` type)?

j0sh replied:
Yeah, an interface is not really necessary here. Removed in b70f22f26f926415ae39a887ca1d74eb7646c1cd

j0sh replied:
The linter was complaining about "stuttering" (`verification.VerificationResult`), and suggested just `Result` so I went with that (`Results` actually), the brevity is much better. Also took the liberty of renaming `VerifierParams` to simply `Params` as well for consistency. https://github.com/livepeer/go-livepeer/pull/1096/commits/cbe81d9208d5f65c63e9914c78781156747c9cc0

yondonfu commented on verification/epic.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+func epicResultsToVerificationResults(er *epicResults) (*verificationResult, error) {
+	// find average of scores and build list of pixels
```
Ah so an implication of our current transcoding workflow where we either receive all renditions requested or we consider the segment transcoding as failed is that if 1 out of N renditions is classified as tampered we retry for all N renditions. I feel like ideally in that case we only retry for the single rendition, but I get why retrying for all N renditions fits better with our current workflow.

j0sh committed:
epic: Use shared volume for verification.

Temporary; most of this should be removed when [1] is implemented.

[1] https://github.com/livepeer/verification-classifier/issues/64

j0sh committed:
server: Perform playlist insertion after verification.

Unclear whether this is the best thing to do unconditionally.

j0sh commented:
Functionally complete but missing tests. There are two ways to get data into the verifier:

- External object storage
- Shared volume

The shared volume (assumed to be mounted in the verifier container as `/stream`) actually works rather nicely for remote verifiers as long as the volume is accessible over the network to go-livepeer. I've been using sshfs with success.

Remote shared volumes along with support for broadcaster-external object storage (which also works nicely; tested with GCS) makes https://github.com/livepeer/verification-classifier/issues/64 a lot less urgent now.

@yondonfu I'd really like to clean up the commit log; let me know when you're OK with a rebase.

kyriediculous commented:
Will give this a look further in the week but I just had two minor questions for now: 

Will this verifier API use a trained model made available to the node so that it only has to run input through an already trained model, meaning that it could be run on lower end machines ? 

What is the default we are aiming for? Running locally or remote?

yondonfu commented on server/broadcast.go:
```diff
@@ -412,14 +423,25 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string)
 			if monitor.Enabled {
 				monitor.TranscodedSegmentAppeared(nonce, seg.SeqNo, sess.Profiles[i].Name)
 			}
+			/* Should we do this conditionally in the presence of the verifier? Otherwise playlist insertions are blocked until all segments are available, which seems worse - but is necessary for verification
```
Seems reasonable to insert into the playlist when segments are available if `verifier = nil`

j0sh replied:
Thought about this and I'd really prefer not to have two code paths that do the same thing depending on which mode the node is running in. That can get very confusing and error-prone very quickly, and I would really like to avoid setting that precedent here.

While we do have a lot of different operating modes on the node, the composition has been rather clear-cut so far: a bit of additional logic might execute for on-chain mode, or an orchestrator may invoke a remote transcoder in standalone mode, but there aren't really cases AFAIK where we execute the same logic in different places, depending on the mode. So that's a slope I'd like to avoid going down as much as possible.

Removed this bit here: https://github.com/livepeer/go-livepeer/pull/1096/commits/55a8d8b683a0dde032081be9201ddae5113f6aa7

yondonfu commented on server/broadcast.go:
```diff
@@ -434,6 +456,23 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string)
+			if err != nil {
+				// We assume playlist insertion is *not* recoverable for now
+				glog.Error("Playlist insertion error; is this recoverable? nonce=%d manifestID=%s seqNo=%d err=%s", nonce, cxn.mid, seg.SeqNo, err)
+				return nil
```
Should we log this error using the `monitor` package similar to what we did when we [inserted into the playlist as segments became available](https://github.com/livepeer/go-livepeer/pull/1096/files#diff-cecc27bfe2b6b19181ce443733dfba57R441)?

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer/pull/1096/commits/a153b1f8adc0725ef2aad4364960e4f40b32a876

yondonfu commented on server/broadcast.go:
```diff
@@ -462,6 +501,85 @@ func shouldStopSession(err error) bool {
+		// Check if an earlier verification attempt was the one accepted.
+		// If so, reset the local OS if we're using that since it's been
+		// overwritten with this rendition.
+		if accepted != params && !sess.BroadcasterOS.IsExternal() {
```
The renditions wouldn't be overwritten in an external OS for multiple verification attempts?

j0sh replied:
If an external OS is used, then each orchestrator writes to a different path. This avoids issues around overwriting, front-running, etc. The actual URL which gets inserted into the playlist is the one that passes verification.

https://github.com/livepeer/go-livepeer/blob/775a76ba16fc71bf901030f994b413f682ff3fdd/server/broadcast.go#L215-L220

yondonfu commented on verification/verify.go:
```diff
@@ -0,0 +1,139 @@
+			return nil, nil
+		}
+		sort.Sort(byResScore(sv.results))
+		return sv.results[0].params, err
```
Given the current definition of [Less()](https://github.com/livepeer/go-livepeer/pull/1096/files#diff-98d66d2dc5372dedc76d5fdde7bedebcR81) won't the first element have the lowest score? Think we would want the last element here to get the result with the highest score

j0sh replied:
Yep - fixed (and tested!)

yondonfu commented:
@j0sh 

> I'd really like to clean up the commit log; let me know when you're OK with a rebase.

👍 to do a rebase. I left a few comments, but feel free to rebase before/after addressing them without a thumbs up. Once you feel like the PR is ready for a full review (I know you mentioned you wanted to add some unit tests and implement the fatal/verifier errors) let me know! In the meantime, I'll do some testing with both the shared Docker volume and an external OS for B.

@kyriediculous 

> Will this verifier API use a trained model made available to the node so that it only has to run input through an already trained model, meaning that it could be run on lower end machines ?

The verifier API would use a trained model.

> What is the default we are aiming for? Running locally or remote?

The user will have to run the verifier API separately from the node so it is up to them whether they want to run the verifier API locally or remotely [1]. For performance, they'll probably want to run it remotely. For ease use of use, we can consider automating the verifier API setup either via a script or via a docker-compose file for Docker users - we can discuss that separately though.

[1] At the moment, this PR enables slightly limited remote verifier API usage - see [this comment](https://github.com/livepeer/go-livepeer/pull/1096#issuecomment-562837768)

j0sh commented:
* Added a number of unit tests, bug fixes resulting from those, etc.
* Addressed various bits of feedback
* Only major piece that doesn't have test coverage is the epic classifier integration code
* Still don't distinguish between retryable and fatal errors per https://github.com/livepeer/go-livepeer/pull/1096#discussion_r355035563

j0sh commented:
>>Will this verifier API use a trained model made available to the node so that it only has to run input through an already trained model, meaning that it could be run on lower end machines ?
>
> The verifier API would use a trained model.

Relatedly, would be good to resolve https://github.com/livepeer/verification-classifier/issues/63 so the node doesn't need to be aware of the model that's being used.

yondonfu commented on server/broadcast.go:
```diff
@@ -432,6 +448,26 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string)
+			err = cpl.InsertHLSSegment(&sess.Profiles[i], seg.SeqNo, url, seg.Duration)
+			if err != nil {
+				// We assume playlist insertion is *not* recoverable for now
+				glog.Errorf("Playlist insertion error; is this recoverable? nonce=%d manifestID=%s seqNo=%d err=%s", nonce, cxn.mid, seg.SeqNo, err)
```
Should we omit "is this recoverable?" from the error message?

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer/pull/1096/commits/b2c942db6a235b82a286d0a08b826a494c05e57a

yondonfu commented on server/broadcast.go:
```diff
@@ -460,6 +496,86 @@ func shouldStopSession(err error) bool {
+				if err != nil {
+					return err
+				}
+				// We *probably* don't need to reset URI[i] here
```
Why would we need a reset here?

j0sh replied:
Whenever we use an external OS, the URI changes, so it does need to be reset. Fixed this bit in https://github.com/livepeer/go-livepeer/pull/1096/commits/7c067159f47ca94b032d8bfd15b2f8997e46d68b and added tests in https://github.com/livepeer/go-livepeer/pull/1096/commits/788ed40861b2fbbc7fcf384d33d77c533abf0345

yondonfu commented on server/broadcast.go:
```diff
@@ -460,6 +496,86 @@ func shouldStopSession(err error) bool {
+				if len(data) <= 0 {
+					return errors.New("MissingLocalData")
+				}
+				// XXX fix required base path
```
By "required base path", do you mean `<PROFILE_NAME>/<SEG_NO>` ([example](https://github.com/livepeer/go-livepeer/blob/master/server/broadcast.go#L382))?

j0sh replied:
The problematic part is actually the `/stream/<manifestID>` part of the base path. The OS `SaveData` method does this:

```
name, err := os.SaveData("<profile>/<segNo>", data)
```

where `name` is a full(er) path - if using a local OS, it'll look something like `/stream/<manifestID>/<profile>/<seqNo>`

However, the returned `name` is not composable with `SaveData()` itself; it can't be passed back into `SaveData` to over-write existing data. Ideally we could do something like this:

```
name1 = os.SaveData("<profile>/<segNo>", "abc")
name2 = os.SaveData(name1, "def")
name1 == name2
os.GetData(name1) == "def"
```

Ideally the OS itself would do the prefix stripping but that needs to be done on a per-OS basis (since the implementation of `SaveData` is per-OS). Right now we only need it for local OS though. We could also strip the prefix within the `verify` function just prior to saving within the OS, but that's an implementation detail leaking out; the driver API is smelly and could use some deeper thought.

Anyway, worked around the issue within `verify` here https://github.com/livepeer/go-livepeer/pull/1096/commits/7c067159f47ca94b032d8bfd15b2f8997e46d68b and tests modified to match / pass in https://github.com/livepeer/go-livepeer/pull/1096/commits/788ed40861b2fbbc7fcf384d33d77c533abf0345

yondonfu commented on server/broadcast_test.go:
```diff
@@ -519,3 +545,212 @@ func TestVerifyPixels(t *testing.T) {
+	assert.NotNil(err)
+	assert.Equal(2, sv.calls)
+	assert.Equal(sv.err, err)
+	assert.Len(bsm.sessMap, 0) // No effect on map for now
```
Should this comment be removed since the length of `bsm.sessMap` decreases for retryable errors such as `verification.ErrTampered`?

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer/pull/1096/commits/788ed40861b2fbbc7fcf384d33d77c533abf0345

yondonfu commented on server/broadcast_test.go:
```diff
@@ -519,3 +545,212 @@ func TestVerifyPixels(t *testing.T) {
+	assert.Equal(sv.err, err)
+	assert.Len(bsm.sessMap, 1) // No effect on map for now
+
+	// Check non-retryable errors, esp broadcast session removal from manager
```
Should this say retryable errors?

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer/pull/1096/commits/788ed40861b2fbbc7fcf384d33d77c533abf0345

j0sh commented:
* Updated PR to incorporate some lingering issues that were commented in the code 
* Addressed remaining PR feedback
* Still needs additional tests for epic classifier
* Still needs to distinguish retryable from fatal errors

There is a bug where the pixels are not checked if the error is retryable. Partial mitigation in https://github.com/livepeer/go-livepeer/pull/1096/commits/f061a0f801359209d2c80b653baf610c950ca505 , tests in https://github.com/livepeer/go-livepeer/pull/1096/commits/5724313a8ee39bb30c0164c723ebb606153a9bea . To really address this correctly, I think we need to distinguish retryable from fatal errors. (A fatal error, such as with pixel counts, would still lead to a retry, but the O is removed from the working set and its results do not have the possibility of being selected.)

j0sh commented:
Added unit tests for the epic classifier. Mostly happy with the state of the tests. There are a few code paths that haven't been exercised, as noted in the comments. The unexercised code paths are mostly around error handling for library functions such as `ioutil.WriteFile`. I'm OK with leaving the tests as-is unless there is a preference to complete those prior to merging. https://github.com/livepeer/go-livepeer/pull/1096/commits/57350e46c577b65d6a5a0d1cb68f06558c7c38a0

Fixes as a result of these tests in https://github.com/livepeer/go-livepeer/pull/1096/commits/f71b4974d2349b3f77068e20166a7f8b8962cc9b and https://github.com/livepeer/go-livepeer/pull/1096/commits/8c030374ecf1d3d839d00154dc681b00ea403027

yondonfu commented on server/broadcast.go:
```diff
@@ -432,6 +448,26 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string)
+		if verifier != nil {
+			err := verify(verifier, cxn, sess, seg, res, segURLs)
+			if err != nil {
+				return err
```
Can we log the error here? [1] I don't see a log statement to handle some of the errors that would be propagated from `verify()`.

[1] An alternative could be to log the error returned by [transcodeSegment()](https://github.com/livepeer/go-livepeer/pull/1096/files#diff-cecc27bfe2b6b19181ce443733dfba57R286)

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer/pull/1096/commits/33856c9130973b277297d0c1032550bebc83d20c

yondonfu commented on server/broadcast_test.go:
```diff
@@ -519,3 +547,290 @@ func TestVerifyPixels(t *testing.T) {
+	// Ensure that the playlist has the correct URL after verification
+	// Following this sequence:
+	//   1. Verify seg1. Verification fails.
+	//   2. Verify seg2. Verification fails.
```
Should there be a step 3:

`3. Verify seg3. Verification fails`

since the test runs verification 3 times?

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer/pull/1096/commits/a908c149a95a16d4add837e2a3fe4bb6b0749716

yondonfu commented on server/broadcast_test.go:
```diff
@@ -519,3 +547,290 @@ func TestVerifyPixels(t *testing.T) {
+	//   1. Verify seg1. Verification fails.
+	//   2. Verify seg2. Verification fails.
+	//   3. Max retries hit, assume false positives
+	//   4. Return seg1 as highest scoring result
```
Should this step and the next step say `seg2` since the test asserts that the second segment is used since it has the highest score?

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer/pull/1096/commits/a908c149a95a16d4add837e2a3fe4bb6b0749716

yondonfu commented on server/broadcast_test.go:
```diff
@@ -519,3 +547,290 @@ func TestVerifyPixels(t *testing.T) {
+		})
+		assert.Nil(err, fmt.Sprintf("Could not marshal results for %s", url))
+		ts, mux := stubTLSServer()
+		//defer ts.Close()
```
Perhaps we can return `ts` from this function and then call `Close()` on each instance of `httptest.Server` that is created via a call to `genBcastSess()`. Something like:

```
genBcastSess := func(url string) (*BroadcastSession, *httptest.Server) { ... }

sess1, ts1 := genBcastSess(...)
sess2, ts2 := genBcastSess(...)
sess3, ts3 := genBcastSess(...)

defer func() {
    ts1.Close()
    ts2.Close()
    ts3.Close()
}()
```

j0sh replied:
Took different approach in https://github.com/livepeer/go-livepeer/pull/1096/commits/a908c149a95a16d4add837e2a3fe4bb6b0749716 . It's a bit of a kludge but the API is more satisfying. Do you still prefer the explicit approach?

Do you know how important it is to call `ts.Close()` for httptest servers?  It seems designed to incorporate server shutdowns into testing rather than any critical bookkeeping [1], but I'm having a hard time finding more detailed explanations of its intended usage.

[1] https://golang.org/src/net/http/httptest/server.go?s=5205:5229#L180

yondonfu replied:
> It's a bit of a kludge but the API is more satisfying. Do you still prefer the explicit approach?

I'm fine with the approach in a908c14 - just felt like it would be good practice to be in the habit of closing/shutting things down when they're no longer used even in tests.

> Do you know how important it is to call ts.Close() for httptest servers? It seems designed to incorporate server shutdowns into testing rather than any critical bookkeeping

Not sure exactly how important it is - the [docs for NewServer()](https://golang.org/pkg/net/http/httptest/#NewServer) do mention that the caller should call Close() when finished. Perhaps its just to make sure that there are no idle connections lying around in tests.

yondonfu commented:
> To really address this correctly, I think we need to distinguish retryable from fatal errors. (A fatal error, such as with pixel counts, would still lead to a retry, but the O is removed from the working set and its results do not have the possibility of being selected.)

And right now we're removing the O from the working set for both verifier (i.e. tampered) and fatal errors (i.e. pixel mismatch, audio mismatch) (which are all categorized as `Retryable` at the moment) i.e. `ErrPixelMismatch`, `ErrTampered`, `ErrAudioMismatch`? That seems ok to me for now.

If we distinguish verifier (i.e. `ErrTampered`) from fatal (i.e. `ErrPixelMismatch` and `ErrAudioMismatch`) separately from this PR then we could continue to always remove O from the working set for both types of errors (since we want to make sure not to retry with the same O), but for fatal errors we could apply a more severe penalty based on a [ban policy](https://github.com/livepeer/go-livepeer/issues/1100#issuecomment-558846820).

> I'm OK with leaving the tests as-is

👍

yondonfu reviewed (commented):
Changes look good to me - let's rebase to clean up the commits and resolve the merge conflict?

kyriediculous commented on cmd/livepeer/livepeer.go:
```diff
@@ -663,6 +667,21 @@ func main() {
 		}
+
+		// Set up verifier
+		if *verifierAddr != "" {
```
For other external URLs entered as flags we prefix the URL with `https://` if it doesn't have a http protocol prefix, do you think it would be a good idea to add it here as well, just to have a uniform approach. Otherwise it might lead to confusion when some flags require the prefix and some don't.

j0sh replied:
This probably needs to be renamed to `verifierUrl` or something [1], including the CLI flag. Most of our other `*Addr` flags technically only take a `host:port` combination, but this is more of a full URL, including the scheme and path such as `http://host:port/path/to/verifier`

[1] For our other external callback URLs (`orchWebhookUrl` and `authWebhookUrl`), we actually fail out if the prefix isn't http/https, rather than trying to correct it. Probably should do similar here - in fact, the below could use some refactoring.

https://github.com/livepeer/go-livepeer/blob/7f441276ac75c235e6578157e462a7ec5d82e275/cmd/livepeer/livepeer.go#L772-L800

kyriediculous replied:
failing when the prefix isn't included sounds good to me !

j0sh replied:
Made the following changes [1]

* Refactored to share webhook URL parsing code (should be no functional changes) 8e9062d393ad69b38b232e3e54179540d992f022

* Updated to use `-verifierUrl` as the flag name, following the scheme (and capitalization 😦) of our external webhook URLs in e62923951858d966623c3e855055039535b18f6d

* Additional CLI tests in b5665a81609ee648de6796680acd76506b1faa6e

[1] Went ahead and fixed up the commits locally rather than pushing a separate fixup commit for review since the timeline here is tight and there would have been merge conflicts due to a few commits changing the same area of code in `livepeer.go`

kyriediculous replied:
Thanks ! Changes look good 👍

j0sh committed:
cmd: Reuse URL parsing code for webhooks.

j0sh commented:
Rebased on top of master, collapsed all the fixups. There were a number of merge conflicts and build-breaking changes. Most of the commits should still compile OK on their own, but I didn't test every single one of them 🤞

yondonfu commented on README.md:
```diff
@@ -128,6 +128,14 @@ A standalone transcoder can be run which connects to a remote orchestrator. The
 
+### Verification
+
+Experimental verification of video using the Epic Labs classifier can be enabled with the `-verifierAddr` flag. Pass in the address of the verifier API:
```
`-verifierAddr` -> `-verifierUrl`?

j0sh replied:
Fixed in 796ba8080309dc3e98a955daccbe46fd2f96845c

yondonfu commented on README.md:
```diff
@@ -128,6 +128,14 @@ A standalone transcoder can be run which connects to a remote orchestrator. The
+
+Experimental verification of video using the Epic Labs classifier can be enabled with the `-verifierAddr` flag. Pass in the address of the verifier API:
+
+- `livepeer -broadcaster -verifierAddr http://localhost:5000/verify`
```
Realized that this command would be insufficient for a user to enable verification at the moment since the broadcaster would either need to also use an external storage bucket or provide the `-verifierPath` flag. Should we add a note about that here? Since livepeer/docs#50 adds a verification section to the RTD we can probably go light on details here.

j0sh replied:
Fixed in 796ba8080309dc3e98a955daccbe46fd2f96845c

yondonfu reviewed (commented):
Left two minor comments about the verification section in the README. Feel free to address them & fixup the relevant commits locally without pushing up separate fixup commits.

j0sh commented:
For some reason the verifier / epic tests fell out of git while rebasing, so re-committed those and re-pushed (luckily does not look there was any data loss).

Here's a demonstration of the problem around retryable errors and pixel counts. Basically, what happens is:

1. O returns a legitimate result with a high score
2. The same results have an incorrect pixel count
3. Verifier detects both the high score and the pixel mismatch
4. Because pixel mismatch errors are marked as retryable, the O with the high score may have its results selected even if its reported pixel count is incorrect.

This is problematic because relying on O's self-reported results could lead B to over-count the pixels that were actually processed, resulting in higher effective fees. If the pixel counts don't match, then:

1. The results should not be eligible for selection
2. B should retry the transcoding

I've been referring to errors that should follow these two behaviors "fatal errors" but I think we need a better term for that. Non-fatal errors (ones that are simply retryable) should also re-add any evicted Os back into the pool, especially if we attempt to distinguish between false negatives and true positives via retries (false negatives re-add, true positives should stay evicted and/or be further penalized with a ban policy.). However, re-adding evicted Os isn't strictly needed to resolve the current issue around the fatal/retryable split.

```diff
diff --git a/verification/verify_test.go b/verification/verify_test.go
index b506e713..2d02ce95 100644
--- a/verification/verify_test.go
+++ b/verification/verify_test.go
@@ -115,6 +115,11 @@ func TestVerify(t *testing.T) {
        res, err = sv.Verify(&Params{ManifestID: "stu", Results: data})
        assert.Equal(ErrPixelsAbsent, err)
        assert.Equal("mno", string(res.ManifestID)) // Still return best result
+       // Higher score and incorrect pixel score should prioritize pixel count
+       verifier.results = &Results{Score: 20.0, Pixels: []int64{123, 123}}
+       res, err = sv.Verify(&Params{ManifestID: "vwx", Results: data})
+       assert.Equal(ErrPixelMismatch, err)
+       assert.Equal("mno", string(res.ManifestID)) // This should not fail! But it does.
 ```

yondonfu reviewed (approved):
LGTM - as discussed in Discord we can address the pixel counting + verification issue in a separate PR.

While this is still experimental and the verifier needs improvement, excited to see this initial integration working! 🎉
